hudi-sync/hudi-hive-sync/src/main/java/org/apache/hudi/hive/util/PartitionFilterGenerator.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.hive.util;
+
+import org.apache.hudi.common.util.ReflectionUtils;
+import org.apache.hudi.hive.HiveSyncConfig;
+import org.apache.hudi.hive.HoodieHiveSyncException;
+import org.apache.hudi.hive.expression.BinaryOperator;
+import org.apache.hudi.hive.expression.Expression;
+import org.apache.hudi.hive.expression.LeafExpression;
+import org.apache.hudi.sync.common.model.FieldSchema;
+import org.apache.hudi.sync.common.model.Partition;
+import org.apache.hudi.sync.common.model.PartitionValueExtractor;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.hudi.hive.HiveSyncConfig.HIVE_SYNC_FILTER_PUSHDOWN_MAX_SIZE;
+import static org.apache.hudi.sync.common.HoodieSyncConfig.META_SYNC_PARTITION_EXTRACTOR_CLASS;
+
+public class PartitionFilterGenerator {
+
+  private interface ExpressionBuilder<T> {
```

Comment:
I think the idea is actually a little bit different: we need visitors to translate Expression tree into a Filter (string), but we don't need it to gen Expression from Partition (for that we can just have a simple method).

To exemplify: we should have something like following:

```
class Expression {
  abstract T accept(ExpressionVisitor<T> v);
}

class BinaryExpression {
  T accept(ExpressionVisitor<T> v) {
     v.visitBinaryExpr(this);
  } 
} 

class ExpressionVisitor<T> {
  T visitBinaryExpr(BinaryExpression binExpr);
  // ...
}

class FilterGenVisitor extends ExpressionVisitor<String> { /* ... */ }

```

Reply:
Will change. Also, I think maybe we still need to keep ExpressionBuilder as well? As it's easy and clear to build the expression in this way.

Reply:
`ExpressionBuilder` doesn't seem to add a lot of value IMO
 - It has just one method (another could be replaced w/ a simple `stream().reduce(...)`
 - There's no polymorphism require where it's used